src/wasm.h

```diff
@@ -583,15 +583,15 @@ class Literal {
   }
   Literal rotL(const Literal& other) const {
     switch (type) {
-      case WasmType::i32: return Literal(RotateLeft(uint32_t(i32), uint32_t(other.i32)));
-      case WasmType::i64: return Literal(RotateLeft(uint64_t(i64), uint64_t(other.i64)));
+      case WasmType::i32: return Literal(RotateLeft(uint32_t(i32), other.i32));
```

Comment:
RotateLeft and RotateRight get compiler warnings because they negate an unsigned value (-count). So we have a compiler warning that needs to go away. I thought I did that in a way that wouldn't change semantics and might be an instruction faster.
http://stackoverflow.com/questions/4267194/compiler-warning-for-unary-operation-on-unsigned-int/4267258#4267258

There are multiple issues involved and I am researching further. Here's some thoughts.
1) What is the intention of the negation and what is the C++ compiler actually doing? I believe the intent was to get a signed value. However, the result of the negation is an unsigned value. You can remove the warning by coding -(int)count or -(int64)count. That might be the compiler's default handling? Looks like typing is i32->u32->int.
2) For sure if my PR changed the semantics, that would be wrong of me.
3) You say the 2 args should be the same type. Why?
4) For sure if my PR added cycles, that would be wrong of me. The ultimate resolution should be same cycles or faster.
5) What is the specification of shift count? Must count be positive or can it also be negative. What happens if it's greater than the number of bits of the left hand operand type?
I'm continuing to look into these questions. What can you tell me?

    On Wednesday, May 11, 2016 9:04 PM, Derek Schuff <notifications@github.com> wrote:
 

 In src/wasm.h:> @@ -583,15 +583,15 @@ class Literal {
>    }
>    Literal rotL(const Literal& other) const {
>      switch (type) {
> -      case WasmType::i32: return Literal(RotateLeft(uint32_t(i32), uint32_t(other.i32)));
> -      case WasmType::i64: return Literal(RotateLeft(uint64_t(i64), uint64_t(other.i64)));
> +      case WasmType::i32: return Literal(RotateLeft(uint32_t(i32), other.i32));
RotateLeft is a template where the 2 arguments are the same type. How does changing this from an explicit cast to an implicit conversion help?—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub